### PR TITLE
feat: Support installing poetry dependencies with pip

### PR DESCRIPTION
**Pull request by pdecat**
## Description
This PR adds support for installing poetry dependencies with `pip --no-deps` for projects with `pyproject.toml` and optionally `poetry.lock` files.

We use pip as as poetry [does not currently provide a proper way to isolate installed dependencies from virtualenv](https://github.com/python-poetry/poetry/issues/1937) (pip's `--target` option). `--no-deps` [disables pip's dependency resolver as poetry already pinned all dependencies](https://github.com/python-poetry/poetry/issues/3472#issuecomment-973170415).

This is a continuation of https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/186

## Motivation and Context
poetry's is more and more used and has a robust dependency resolver. 

## Breaking Changes
New poetry install step is not enabled by default and requires explicitly setting `poetry_install = True` in the source block.

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [x] I have executed `pre-commit run -a` on my pull request

```
cd examples/build-package
terraform init
rm -rf builds/lambda_layer/ && terraform apply -target module.lambda_layer -auto-approve
rm -rf builds/lambda_layer_poetry/ && terraform apply -target module.lambda_layer_poetry -auto-approve
rm -rf builds/package_dir/ && terraform apply -target module.package_dir -auto-approve
rm -rf builds/package_dir_poetry/ && terraform apply -target module.package_dir_poetry -auto-approve
terraform destroy
```

### CODE REVIEW

File: examples/build-package/main.tf [hunk 1]
```diff
@@ -278,8 +302,31 @@ module "lambda_layer" {
 
   build_in_docker = true
   runtime         = "python3.8"
-  docker_image    = "public.ecr.aws/sam/build-python3.8"
+  docker_image    = "build-python3.8"
```
Comment (pdecat) [inline examples/build-package/main.tf:305]: If `public.ecr.aws/sam/build-python3.8` already exists locally, the `docker_file` is not used.

File: package.py [hunk 2]
```diff
@@ -729,6 +756,7 @@ def commands_step(path, commands):
                 if runtime.startswith('python'):
                     pip_requirements_step(
                         os.path.join(path, 'requirements.txt'))
```
Comment (pdecat) [inline package.py:758]: `pip_requirements_step()` already appends the filename if `path` is a directory: 
```suggestion
                    pip_requirements_step(path)
```

File: examples/build-package/main.tf [hunk 3]
```diff
@@ -17,28 +17,52 @@ resource "random_pet" "this" {
...
+
+  build_in_docker = true
+  runtime         = "python3.8"
+  docker_image    = "build-python3.8-poetry"
+  docker_file     = "${path.module}/../fixtures/python3.8-app-poetry/docker/Dockerfile"
+
+  source_path = [
+    {
+      path           = "${path.module}/../fixtures/python3.8-app-poetry"
+      poetry_install = true
```
Comment (pdecat) [inline examples/build-package/main.tf:62]: Maybe rename this to be more explicit that `poetry export` then `pip install --no-deps` are executed.

**Comment by pdecat** [comment]
Still current

**Comment by pdecat** [comment]
Still current

**Comment by 1davidmichael** [comment]
I primarily use poetry with Python and appreciate this PR. I hope it is merged at some point in the future. Thanks for this work!

**Comment by antonbabenko** [comment]
@pdecat Please let us know if/when this PR is ready for review and merge?

I expect that other users (including @1davidmichael @panessa @Tonkonozhenko) will be able to help with the review and testing.

**Comment by pdecat** [comment]
Hi, I believe it is ready for review and testing by others.

File: package.py [hunk 4]
```diff
@@ -967,6 +1007,122 @@ def install_pip_requirements(query, requirements_file, tmp_dir):
...
+                        poetry_cache_dir=poetry_cache_dir,
+                    )
+                )
+            else:
+                cmd_log.info(poetry_commands)
+                log_handler and log_handler.flush()
+                for poetry_command in poetry_commands:
+                    check_call(poetry_command, env=subproc_env)
+
+            os.remove(poetry_lock_target_file)
```
Comment (panessa) [inline package.py:1120]:
```python
            if poetry_toml_target_file:
                os.remove(poetry_toml_target_file)
```

File: package.py [hunk 5]
```diff
@@ -967,6 +1007,122 @@ def install_pip_requirements(query, requirements_file, tmp_dir):
...
+
+        # Install dependencies into the temporary directory.
+        with cd(temp_dir):
+            # NOTE: poetry must be available, which is the case with lambci/lambda:build-python* docker images
+            # FIXME: poetry install does not currently allow to specify the target directory so we export the 
+            # requirements then install them with pip --no-deps to avoid using pip dependency resolver
+            poetry_commands = [
+                shlex_join([ poetry_exec, "config", "--no-interaction", "virtualenvs.create", "true" ]),
+                shlex_join([ poetry_exec, "config", "--no-interaction", "virtualenvs.in-project", "true" ]),
+                shlex_join([ poetry_exec, "export", "-f", "requirements.txt", "--output", "requirements.txt" ]),
```
Comment (panessa) [inline package.py:1084]: Replace with
```python
shlex_join([ poetry_exec, "export", "--with-credentials", "-f", "requirements.txt", "--output", "requirements.txt" ]),
```

File: package.py [hunk 6]
```diff
@@ -967,6 +1007,122 @@ def install_pip_requirements(query, requirements_file, tmp_dir):
...
+
+    log.info("Installing python dependencies with poetry & pip: %s", poetry_lock_file)
+    with tempdir() as temp_dir:
+        def copy_file_to_target(file, temp_dir):
+            filename = os.path.basename(file)
+            target_file = os.path.join(temp_dir, filename)
+            shutil.copyfile(file, target_file)
+            return target_file
+
+        poetry_lock_target_file = copy_file_to_target(poetry_lock_file, temp_dir)
```
Comment (panessa) [inline package.py:1065]:
```python
        if os.path.isfile(poetry_toml_file):
            poetry_toml_target_file = copy_file_to_target(poetry_toml_file, temp_dir)
        else:
            poetry_toml_target_file = None
```

File: package.py [hunk 7]
```diff
@@ -967,6 +1007,122 @@ def install_pip_requirements(query, requirements_file, tmp_dir):
             yield temp_dir
 
 
+@contextmanager
+def install_poetry_dependencies(query, path):
+    # TODO:
+    #  1. Emit files instead of temp_dir
+    poetry_lock_file = os.path.join(path, "poetry.lock")
```
Comment (panessa) [inline package.py:1014]:
```python
poetry_toml_file = os.path.join(path, "poetry.toml")
```

File: package.py [hunk 8]
```diff
@@ -660,6 +673,20 @@ def pip_requirements_step(path, prefix=None, required=False, tmp_dir=None):
...
+            if get_build_system_from_pyproject_toml(pyproject_file) != "poetry":
+                if required:
+                    raise RuntimeError("poetry configuration not found: {}".format(pyproject_file))
+            else:
+                step("poetry", runtime, path, prefix)
+                hash(pyproject_file)
+                poetry_lock_file = os.path.join(path, "poetry.lock")
+                if os.path.isfile(poetry_lock_file):
+                    hash(poetry_lock_file)
+
```
Comment (panessa) [inline package.py:689]: Include poetry.toml for http-basic auth
```python
                poetry_toml_file = os.path.join(path, "poetry.toml")
                if os.path.isfile(poetry_toml_file):
                    hash(poetry_toml_file)
```

**Review by panessa — COMMENTED** [review]
I added support for http-basic auth for poetry. Code was not taking in consideration poetry.toml file

File: examples/build-package/main.tf [hunk 9]
```diff
@@ -17,28 +17,52 @@ resource "random_pet" "this" {
...
+
+# Create zip-archive of a single directory where "poetry export" & "pip install --no-deps" will also be executed
+module "package_dir_poetry" {
+  source = "../../"
+
+  create_function = false
+
+  build_in_docker = true
+  runtime         = "python3.8"
+  docker_image    = "build-python3.8-poetry"
```
[the remark this answers is not in the document]
Comment (pdecat) [reply examples/build-package/main.tf]: Done, thanks!

**Review by pdecat — COMMENTED** [review]
Hi @panessa, thanks for the review! I've integrated your suggested changes, please take a look.

**Comment by pdecat** [comment]
Hi @antonbabenko, some Github Actions jobs failed because of network errors. Mind to restart them? :pray:

**Comment by antonbabenko** [comment]
@pdecat Restarted GitHub Actions, and all checks are green. I have scheduled to revive this PR on the 13th of October (after HashiConf). Hopefully, everyone else will be addressing all the comments and do a review, too.

File: test_package_toml.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+from package import get_build_system_from_pyproject_toml
```
Comment (antonbabenko) [inline test_package_toml.py:1]: Can we move this file from the root to keep it clean?

Also, could you please update README with some instructions on how to run these? Or just delete it :)

File: examples/fixtures/unittests/pyproject-unknown.toml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+[build-system]
```
Comment (antonbabenko) [inline examples/fixtures/unittests/pyproject-unknown.toml:1]: Do we need this file, too?

File: examples/build-package/main.tf [hunk 12]
```diff
@@ -17,28 +17,52 @@ resource "random_pet" "this" {
...
+
+# Create zip-archive of a single directory where "poetry export" & "pip install --no-deps" will also be executed
+module "package_dir_poetry" {
+  source = "../../"
+
+  create_function = false
+
+  build_in_docker = true
+  runtime         = "python3.8"
+  docker_image    = "build-python3.9-poetry"
```
Comment (antonbabenko) [inline examples/build-package/main.tf:56]: `docker_image` and `runtime` use different versions of Python. Is it intentional?

**Review by antonbabenko — CHANGES_REQUESTED** [review]
I just tried to test it locally after #362 was merged, and apparently, it broke `package.py` you've created.

The error message I got during `terraform destroy` after the successful run of `rm -rf builds/lambda_layer/ && terraform apply -target module.lambda_layer`:
```
│     npm_requirements_step(
│   File ".../terraform-aws-modules/terraform-aws-lambda/examples/build-package/../../package.py", line 709, in npm_requirements_step
│     raise RuntimeError(
│ RuntimeError: Nodejs interpreter version equal to
│ defined lambda runtime (nodejs14.x) should be available
│ in system PATH
│ 
│ State: exit status 1
```

Could you please fix that and update README.md in the root to showcase poetry support in [this section](https://github.com/terraform-aws-modules/terraform-aws-lambda#combine-various-options-for-extreme-flexibility) and maybe also in other places where it makes sense?

File: test_package_toml.py (hunk 10 above)
Comment (pdecat) [reply test_package_toml.py]: Moved python unit tests to `tests/` folder, add instructions to README.

And also added Github Action to run these tests against all supported python versions.

File: examples/fixtures/unittests/pyproject-unknown.toml (hunk 11 above)
Comment (pdecat) [reply examples/fixtures/unittests/pyproject-unknown.toml]: Yes, it is needed by python unit tests. Moved it too.

File: examples/build-package/main.tf (hunk 12 above)
Comment (pdecat) [reply examples/build-package/main.tf:56]: Nope :) Fixed.

**Comment by pdecat** [comment]
> I just tried to test it locally after #362 was merged, and apparently, it broke `package.py` you've created.
> 
> The error message I got during `terraform destroy` after the successful run of `rm -rf builds/lambda_layer/ && terraform apply -target module.lambda_layer`:

The same errors happens using current master branch.

**Comment by pdecat** [comment]
The _Pre-Commit / Max TF pre-commit_ check is failing with a weird error:

```
Terraform validate with tflint...........................................Failed
- hook id: terraform_tflint
- exit code: 1

Command 'tflint --init' successfully done:




TFLint in modules/deploy/:
Failed to check ruleset; Failed to check `terraform_deprecated_index` rule: main.tf:52,8-9: Invalid character; This character is not used within the language., and 57 other diagnostic(s)
```

See https://github.com/terraform-aws-modules/terraform-aws-lambda/actions/runs/3312065827/jobs/5468288824

And does not only affect jobs from this PR, other PRs are also affected, see https://github.com/terraform-aws-modules/terraform-aws-lambda/actions/runs/3313158688/jobs/5470763712

**Comment by pdecat** [comment]
Pre-commit hooks should be fixed now: https://github.com/terraform-linters/tflint/releases/tag/v0.42.2

**Comment by antonbabenko** [comment]
This PR is included in version 4.3.0 :tada:

**Comment by Tonkonozhenko** [comment]
@pdecat @antonbabenko thank you for the work! we'll use it asap

**Comment by antonbabenko** [comment]
This issue has been resolved in version 4.16.0 :tada:

**Pull request merged**